src/utils/spark.py

```diff
@@ -0,0 +1,223 @@
+        database_url: The JDBC URL for the database connection. Constructed
+            from the above attributes.
+        ipts_password: The password for the database, read from file.
+        fetch_size: The fetch size for the database queries. This is a tuning
+            parameter for query speed. ~10,000 seems to work best.
+        initial_compression: The compression type for the initial Parquet files
+            written via JDBC extract. Defaults to snappy.
+        final_compression: The compression type final repartitioned Parquet
+            files. Defaults to ztd.
+        spark: The Spark session object.
+    """
+
+    def __init__(self, app_name: str, password_file_path: str) -> None:
+        self.app_name = app_name
+        self.password_file_path = password_file_path
+
+        # Vars here are loaded from the .env file, then forwarded to the
+        # container in docker-compose.yaml
+        self.ipts_hostname = os.getenv("IPTS_HOSTNAME")
+        self.ipts_port = os.getenv("IPTS_PORT")
+        self.ipts_service_name = os.getenv("IPTS_SERVICE_NAME")
+        self.ipts_username = os.getenv("IPTS_USERNAME")
+        self.database_url = (
+            f"jdbc:oracle:thin:@//{self.ipts_hostname}:"
+            f"{self.ipts_port}/"
+            f"{self.ipts_service_name}"
+        )
+
+        # Load runtime secret using Compose secrets setup
+        with open(self.password_file_path, "r") as file:
+            self.ipts_password = file.read().strip()
+
+        # Static arguments/params applied to jobs using this session
+        self.fetch_size = "10000"
+        self.initial_compression = "snappy"
+        self.final_compression = "zstd"
+
+        self.spark = SparkSession.builder.appName(self.app_name).getOrCreate()
+
+
+class SparkJob:
+    """
+    Class to manage Spark jobs for reading, repartitioning, and uploading data.
+    Each job corresponds to a single iasWorld table.
+
+    Attributes:
+        session: The shared Spark session containing the Spark connection and
+            database credentials.
+        table_name: The name of the iasWorld table to read from. Should be
+            predicated with 'iasworld.'.
+        taxyr: The tax year(s) to filter and partition by.
+        cur: The cur values to filter and partition by.
+        predicates: A list of SQL predicates for chunking JDBC reads.
+        use_partitions: Whether to partition the output data by taxyr and cur.
+        initial_dir: The initial directory to write the data to, relative to
+            the Docker container.
+        final_dir: The final directory to write the repartitioned data to,
+            relative to the Docker container.
+    """
+
+    def __init__(
+        self,
+        session: SharedSparkSession,
+        table_name: str,
+        taxyr: int | list[int] | None,
+        cur: str | list[str] | None,
+        predicates: list[str] | None,
+        use_partitions: bool,
```

Comment:
**[Thought, non-blocking]** Not a big deal since it's one extra param that directly maps onto a job config param, but it seems like the value of `use_partitions` could be derivable from the values of `cur` and `taxyr` in case we want to simplify things by getting rid of a parameter.